File: 2018/day7/python/main.py
import math
import re

def parseInput():
    f = open("../input", "r")
    lines = f.read().splitlines()
    regex = r'Step (.) .* step (.)'
    pattern = re.compile(regex)
    for line in lines:
        match = pattern.search(line)
        parent = match.group(1)
        child = match.group(2)
        if parent in parentChildrenGraph:
            parentChildrenGraph[parent].append(child)
        else:
            parentChildrenGraph[parent] = [child]
        
        if child in childrenParentGraph:
            childrenParentGraph[child].append(parent)
        else:
            childrenParentGraph[child] = [parent]

def allParentsInOrder(order, parents):
    foundAll = True
    for parent in parents:
        if parent not in order:
            foundAll = False
    return foundAll

def addAvailable(parent, order, available):
    if parent not in parentChildrenGraph:
        return available
    for child in parentChildrenGraph[parent]:
        if allParentsInOrder(order, childrenParentGraph[child]):
            available.append(child)
    available = sorted(available)

# Bi-directional relationship of all parents and children
parentChildrenGraph = {} # [parent, [children]]
childrenParentGraph = {} # [child, [parents]]
parseInput()

rootParents = []
for parent in parentChildrenGraph:
    if parent not in childrenParentGraph:
        rootParents.append(parent)

order = ""
available = rootParents
while len(available) != 0:
    order += available[0]
    available = available[1:]
    addAvailable(order[-1], order, available)

print(order)

def work(numWorkers, available):
    workers = [("", 0)] * numWorkers # (currentTask, whenAvailableAgain)
    time = 0
    order = "" # keep track of processed instructions
    while True:
        for i, w in enumerate(workers): # Free workers and add new availables
            if w[0] != "" and w[1] < time:
                order += w[0]
                addAvailable(w[0], order, available)
                workers[i] = ("", 0) # Worker freed up

        removeAvailable = []
        for a in available: # Assign new work if available
            for j, w in enumerate(workers):
                if w[0] == "":
                    workers[j] = (a, time+ord(a)-ord("A")+60)
                    removeAvailable.append(a)
                    break # Work assigned, continue with next available
        
        for remove in removeAvailable:
            i = available.index(remove)
            available = available[:i]+available[i+1:]

        if all(w[0] == "" for w in workers):
            return time
        time += 1

print(work(5, rootParents))

File: 2018/day7/go/src/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"sort"
	"strings"
)

func readFileToLines(filepath string) (lines []string) {
	file, err := os.Open(filepath)
	defer file.Close()
	if err != nil {
		log.Fatal(err)
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return
}

func parseInstructions(instrList []string) (children map[string][]string, parents map[string][]string) {
	children = make(map[string][]string)
	parents = make(map[string][]string)
	p := "Step (.) .* step (.)"
	re := regexp.MustCompile(p)
	for _, instr := range instrList {
		match := re.FindStringSubmatch(instr)
		lhs := match[1]
		rhs := match[2]
		children[lhs] = append(children[lhs], rhs)
		parents[rhs] = append(parents[rhs], lhs)
	}
	return
}

func determineRootParents(children map[string][]string, parents map[string][]string) (rootParents []string) {
	for parent := range children {
		if _, ok := parents[parent]; !ok {
			rootParents = append(rootParents, parent)
		}
	}
	sort.Strings(rootParents)
	return
}

func determineWorkOrder(inputFilepath string) (order string) {
	lines := readFileToLines(inputFilepath)
	children, parents := parseInstructions(lines)
	available := determineRootParents(children, parents)
	for len(available) != 0 {
		order += available[0]
		available = available[1:]
		for _, new := range children[string(order[len(order)-1])] {
			allParentsInOrder := true
			for _, parent := range parents[new] {
				if !strings.Contains(order, parent) {
					allParentsInOrder = false
				}
			}
			if allParentsInOrder {
				available = append(available, new)
			}
		}
		sort.Strings(available)
	}
	fmt.Printf("Theoretical order: %s\n", order)
	return
}

func all(vs []worker, f func(worker) bool) bool {
	for _, v := range vs {
		if !f(v) {
			return false
		}
	}
	return true
}

type worker struct {
	currInstr int
	doneTime  int
}

func workOrder(inputFilepath string, numWorkers int, workTime int) (time int) {
	lines := readFileToLines(inputFilepath)
	children, parents := parseInstructions(lines)
	available := determineRootParents(children, parents)
	order := ""
	workers := make([]worker, numWorkers)
	for time = 0; ; time++ {
		for i, w := range workers {
			if w.currInstr != 0 && w.doneTime < time {
				order += string(w.currInstr)
				for _, new := range children[string(w.currInstr)] {
					allParentsInOrder := true
					for _, parent := range parents[new] {
						if !strings.Contains(order, parent) {
							allParentsInOrder = false
						}
					}
					if allParentsInOrder {
						available = append(available, new)
					}
				}
				sort.Strings(available)
				workers[i] = worker{0, 0}
			}
		}

		removeBeforeIndex := -1
		for i, a := range available {
			for j, w := range workers {
				if w.currInstr == 0 {
					job := a[0]
					workers[j] = worker{int(job), time + workTime + int(job) - int('A')}
					removeBeforeIndex = i
					break
				}
			}
		}
		available = available[removeBeforeIndex+1:]

		if all(workers, func(w worker) bool {
			return w.currInstr == 0
		}) {
			fmt.Printf("Time: %d\n", time)
			return
		}
	}
}

func main() {
	determineWorkOrder("../input")
	workOrder("../input", 5, 60)
}
